packages/katello/katello/hostname-change.rb

```diff
@@ -239,6 +291,122 @@ def setup_opt_parser
       @opt_parser.parse!
     end
 
+    def dns_managed?
+      @scenario_answers['foreman_proxy']['dns'] &&
+        @scenario_answers['foreman_proxy']['dns_managed'] &&
+        @scenario_answers['foreman_proxy']['dns_provider'] == 'nsupdate'
+    end
+
+    def should_update_dns?
+      dns_managed? && !@options[:skip_dns]
+    end
+
+    def query_dns_server(fqdn, zone, nameserver_ip)
+      query_dns = Resolv::DNS.new(nameserver: [nameserver_ip], search: [], ndots: 1)
+      a_record = query_dns.getresource(fqdn, Resolv::DNS::Resource::IN::A)
+      ip = a_record.address.to_s if a_record
+      soa_record = query_dns.getresource(zone, Resolv::DNS::Resource::IN::SOA)
+      serial = soa_record.serial if soa_record
+      {
+        ip: ip,
+        serial: serial
+      }
+    end
+
+    def new_dns_values
+      new_vals = OpenStruct.new
+
+      new_vals.dns_server = @scenario_answers['foreman_proxy']['dns_server']
+      new_vals.zone = @scenario_answers['foreman_proxy']['dns_zone']
+      new_vals.key_file = @scenario_answers['foreman_proxy']['keyfile']
+
+      new_vals.reverse_zones = [@scenario_answers['foreman_proxy']['dns_reverse']].flatten
+      ip_serial_data = query_dns_server(@old_hostname, new_vals.zone, new_vals.dns_server)
+      new_vals.ip = ip_serial_data[:ip]
+      run_cmd('rndc thaw')
+
+      new_vals
+    end
+
+    def nsupdate_command(dns_entries, key_file)
+      "echo -e \"#{dns_entries}\" | nsupdate -l -k #{key_file}"
+    end
+
+    def update_zone(zone, nameserver_ip)
+      resolver = Resolv::DNS.new(nameserver: [nameserver_ip], search: [], ndots: 1)
+      commands = []
+
+      soa = resolver.getresource(zone, Resolv::DNS::Resource::IN::SOA)
+      commands << "update add #{zone} #{soa.ttl} SOA #{@new_hostname}. #{soa.rname} #{soa.serial + 1} #{soa.refresh} #{soa.retry} #{soa.expire} #{soa.minimum}"
+
+      nameservers = resolver.getresources(zone, Resolv::DNS::Resource::IN::NS)
+      # TODO: replace the correct nameserver
```

Comment:
You are breaking the workflow where a user has multiple nameservers. This is relevant if they set up replication to other nameservers.

Reply:
@ekohl I removed the condition because if we don't update NS records, it fails the post-update name server sanity check and the entire `nsupdate` command errors.  The `update add` and `update delete` commands specify which hostname is applicable, so I don't understand what would be breaking?  What should I do for now and what needs to be in the todo for later?

Reply:
In DNS records are sets. In practice you often only see one record, but it must be treated as a set. You can see this:
```console
$ dig ns example.com

; <<>> DiG 9.11.14-RedHat-9.11.14-2.fc31 <<>> ns example.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27338
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;example.com.			IN	NS

;; ANSWER SECTION:
example.com.		13975	IN	NS	b.iana-servers.net.
example.com.		13975	IN	NS	a.iana-servers.net.

;; Query time: 11 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Fri Mar 06 15:52:55 CET 2020
;; MSG SIZE  rcvd: 88
$ dig ns google.com

; <<>> DiG 9.11.14-RedHat-9.11.14-2.fc31 <<>> ns google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 16103
;; flags: qr rd ra; QUERY: 1, ANSWER: 4, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;google.com.			IN	NS

;; ANSWER SECTION:
google.com.		13667	IN	NS	ns3.google.com.
google.com.		13667	IN	NS	ns4.google.com.
google.com.		13667	IN	NS	ns2.google.com.
google.com.		13667	IN	NS	ns1.google.com.

;; Query time: 10 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Fri Mar 06 15:56:29 CET 2020
;; MSG SIZE  rcvd: 111
```

There's more than 1 nameserver. The way bind replication works is that it will send a notify to all nameservers (besides itself) when there's an update. Assuming all permissions are correct, you can add a second server next to the foreman and have it slave the zone. If delegation is set up correct, when Foreman is down you'll still have your DNS zones available.

What you are doing here is destroying that if the user has set it up by issuing a `delete` and adding back a single record. To avoid this, look at all NS records, see if `old_fqdn` is in there. If so, that should be replaced with `new_fqdn` and add the rest of the set again.

You can argue that this really is an edge case and few users will have set this up, but I like to be correct rather than getting a bug report later that we broke a production setup.

Reply:
@ekohl Thanks for explaining.   The `delete` command I issue is
```ruby
"update delete #{zone}. IN NS #{@old_hostname}"
```

In [`man nsupdate`](https://linux.die.net/man/8/nsupdate) I found

> __update delete domain-name [ ttl ] [ class ] [ type [ data...] ]__
> Deletes any resource records named domain-name. __If type and data is provided, only matching resource records will be removed.__

Since I am providing `type` (NS) and `data` (`@old_hostname` which is the fqdn) in the command, will this address the issue?

Reply:
Yes, I think it does. In that case you can remove the TODO. You can also remove the `namservers = ...` call if you're not going to use the result. I still think the `if` I wrote originally is correct though I used `old_fqdn` and you use `old_hostname`. Perhaps that was the reason it didn't work for you?

Reply:
@ekohl, yeah I agree `old_fqdn` is a better variable name because it makes clear that it's fully qualified.  But earlier in this PR I was convinced by @jturel to just use the existing instance variable `@old_hostname`, since we make sure that's the FQDN.  I think the `if` would work fine, but it's not needed because the specificity of the `delete` command provides the same result.

Reply:
Technically the user could have changed their nameserver to `ns1.example.com` instead of `foreman.example.com`. My version caught that. Same story with the SOA record. However, I' won't push on not catching those since I really doubt anyone who cares about DNS will use the `change-hostname` script.

Reply:
@ekohl Ahh I see what you were saying now.